src/core/tsi/transport_security_interface.h

```diff
@@ -208,76 +209,195 @@ typedef struct {
+    tsi_frame_protector **protector);
+
+/* This method returns the unused bytes from the handshake. It returns TSI_OK
+   assuming there is no fatal error.
+   The caller should not free the bytes.  */
+tsi_result tsi_handshaker_result_get_unused_bytes(
+    const tsi_handshaker_result *self, unsigned char **bytes,
+    size_t *byte_size);
+
+/* This method releases the tsi_handshaker_handshaker object. After this method
+   is called, no other method can be called on the object.  */
+void tsi_handshaker_result_destroy(tsi_handshaker_result *self);
+
 /* --- tsi_handshaker objects ----
 
    Implementations of this object must be thread compatible.
 
-   A typical usage of this object would be:
+   ------------------------------------------------------------------------
+
+   A typical usage of the synchronous TSI handshaker would be:
 
    ------------------------------------------------------------------------
-   tsi_result result = TSI_OK;
+   tsi_result status = TSI_OK;
    unsigned char buf[4096];
-   size_t buf_offset;
-   size_t buf_size;
+   const size_t buf_size = 4906;
+   size_t bytes_received_size = 0;
+   unsigned char *bytes_to_send = NULL;
+   size_t bytes_to_send_size = 0;
+   tsi_handshaker_result *result = NULL;
+
    while (1) {
-     // See if we need to send some bytes to the peer.
-     do {
-       size_t buf_size_to_send = sizeof(buf);
-       result = tsi_handshaker_get_bytes_to_send_to_peer(handshaker, buf,
-                                                         &buf_size_to_send);
-       if (buf_size_to_send > 0) send_bytes_to_peer(buf, buf_size_to_send);
-     } while (result == TSI_INCOMPLETE_DATA);
-     if (result != TSI_OK) return result;
-     if (!tsi_handshaker_is_in_progress(handshaker)) break;
-
-     do {
-       // Read bytes from the peer.
-       buf_size = sizeof(buf);
-       buf_offset = 0;
-       read_bytes_from_peer(buf, &buf_size);
-       if (buf_size == 0) break;
-
-       // Process the bytes from the peer. We have to be careful as these bytes
-       // may contain non-handshake data (protected data). If this is the case,
-       // we will exit from the loop with buf_size > 0.
-       size_t consumed_by_handshaker = buf_size;
-       result = tsi_handshaker_process_bytes_from_peer(
-           handshaker, buf, &consumed_by_handshaker);
-       buf_size -= consumed_by_handshaker;
-       buf_offset += consumed_by_handshaker;
-     } while (result == TSI_INCOMPLETE_DATA);
-
-     if (result != TSI_OK) return result;
-     if (!tsi_handshaker_is_in_progress(handshaker)) break;
+     status = tsi_handshaker_next(
+         handshaker, buf, bytes_received_size,
+         &bytes_to_send, &bytes_to_send_size, &result, NULL, NULL);
+     if (status == TSI_INCOMPLETE_DATA) {
+       // Need more data from the peer.
+       bytes_received_size = buf_size;
+       read_bytes_from_peer(buf, &bytes_received_size);
+       continue;
+     }
+     if (status != TSI_OK) return status;
```

Comment:
I actually would remove this altogether. The code below that handles both cases is clear enough and this sync-only code illustration adds more entropy than real value IMO.